### ISSUE TITLE
A possible bug to lookup credentials

### ISSUE THREAD

**Issue by rochards** [Yubico/java-webauthn-server#138]
Hello,
In classe `InMemoryRegistrationStorage` we have a storage defined as
```java
private final Cache<String, Set<CredentialRegistration>> storage =
        CacheBuilder.newBuilder().maximumSize(1000).expireAfterAccess(1, TimeUnit.DAYS).build();
```
Considering the collection bellow represents one register in `storage`. Some fields have been hidden and others are 
represented as strings for easy viewing:
```json
[
  {
    "key": "parker.peter",
    "credentialRegistration": [
      {
        "signatureCount": 0,
        "userIdentity": {
          "name": "parker.peter",
          "id": "userHandleBytes1"
        },
        "credential": {
          "credentialId": "credentialIdBytes1",
          "userHandle": "userHandleBytes1",
          "publicKeyCose": "publicKeyCoseBytes1",
          "signatureCount": 0
        },
        "attestationMetadata": {}
      },
      {
        "signatureCount": 0,
        "userIdentity": {
          "name": "parker.peter",
          "id": "userHandleBytes2"
        },
        "credential": {
          "credentialId": "credentialIdBytes2",
          "userHandle": "userHandleBytes2",
          "publicKeyCose": "publicKeyCoseBytes2",
          "signatureCount": 0
        },
        "attestationMetadata": {}
      }
    ]
  }
]
```
and the snippet of the `publicKeyCredential` returned from the client as follows:
```json
{
  "id": "credentialIdBytes2",
  "response": {},
  "type": "public-key"
}
```
`response` doesn't have a `userHandle`.

At line 152 of the class `FinishAssertionSteps` we have `credentialRepository.getUserHandleForUsername(request.
getUsername().get())) // request has "parker.peter" as username`. It could bring the first `userHandle` from the 
collection above. So, when it's called `credentialRepository.lookup("credentialIdBytes2", "userHandleBytes1")` at 
line 206 in the same class, it will be empty, throwing the error message `"Unknown credential: credentialIdBytes2"`.

Thank you.

**Comment by emlun** [comment]
Hi! Hm, this is interesting. On the one hand, the [`CredentialRepository`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/1.7.0/com/yubico/webauthn/CredentialRepository.html#getUserHandleForUsername(java.lang.String)) interface declares that usernames should correspond one-to-one with user handles, meaning the above situation (with two user handles for the same username) is in violation of the interface contract. So in that sense it's fine that `FinishAssertionSteps` might produce an incorrect result if the interface is not respected. And although `InMemoryRegistrationStorage` doesn't itself enforce this interface constraint (which is admittedly a code smell), `WebAuthnServer` as implemented will always use the same user handle for the same username.

But on the other hand, the main reason we need to look it up using `getUserHandleForUsername` is because it's needed for `CredentialRepository.lookup(credentialId, userHandle)` - but `lookup` itself returns the user handle in the `RegisteredCredential` result, so that user handle parameter isn't really necessary for anything except maybe for a database index. And `FinishRegistrationSteps` doesn't allow registering multiple credentials with the same credential ID, so it should be safe to remove the `userHandle` parameter from `lookup`. But that would of course be a breaking change, so it's not going to happen in the 1.x version range.

So, in conclusion: not a problem for now, might reconsider the `lookup` method signature in the next major version.

**Comment by rochards** [comment]
Hum... I understand that `username` should correspond one-to-one with `userhandle`. Though, I'm wondering when the user changes his device and wants to register a new credential for that, on the same Relying Party. That way the new device will not know about the credential in the `excludeCredentials` in the `PublicKeyCredentialCreationOptions` and will generate a new one with another credential ID, by-passing the `credentialRepository.lookupAll(response.getId()).isEmpty`. 
Also when we initiate a registration ceremony the `userhandle` is a random sequence byte -> https://github.com/Yubico/java-webauthn-server#registration. Wouldn't we now have the situation described previously? One `username` with two credentials, each one having distinct `userHandle`?

**Comment by emlun** [comment]
Hm, I suppose the README should be clarified to not suggest that the user handle should _always_ be a random byte sequence - rather it should use a new random sequence if the user doesn't already exist, but use the user's existing user handle if it does exist. The demo app works like that too. I'll clarify that in the README.

**Comment by rochards** [comment]
So, it's not a bug at all. Just a misunderstanding from my side.

Thank you for the explanations.

**Comment by emlun** [comment]
Thanks for bringing it up!
